### DISCUSSION TITLE
Fix flaky tests

### DISCUSSION THREAD

**Discussion by colin-grant-work** [eclipse-theia/theia#10790] [Improvements] [answered]
Our CI fails a lot for reasons unrelated to the PR being tested. Usually a few flaky tests are the culprits. We recently disabled the TypeScript code lens test because it was so flaky and had such a long timeout, but there are still a few tests that regularly fail for no particular reason.

 - tests for `closeEditorOnDelete`
 - tests from the `launch-preferences.spec.js` suite. This one in particular runs dozens of permutations of the same test with little informational value, but every run has a chance to miss a timeout.

These failures are pretty annoying for the developer, since we either have to just rerun the tests and hope for green or rebase or make some meaningless change and hope for green, and that eats up time and causes PR's to languish while the CI works itself out.

**Comment by colin-grant-work** [comment]
I'd also add here continued misgivings about the new Playwright tests (@JonasHelming, @planger). À propos [a failure in CI](https://github.com/eclipse-theia/theia/runs/5322296169?check_suite_focus=true), I went to run the tests locally (MacOS) and got numerous failures. I switched to `master` and still got numerous failures; I did a thorough `git clean` and deleted `~/.theia` and got one fewer failures, but still many:

```
[playwright]   Slow test file: ../../tests/theia-quick-command.test.ts (4m)
[playwright]   Slow test file: ../../tests/theia-explorer-view.test.ts (2m)
[playwright]   Slow test file: ../../tests/theia-preference-view.test.ts (2m)
[playwright]   Slow test file: ../../tests/theia-problems-view.test.ts (2m)
[playwright]   Consider splitting slow test files to speed up parallel execution
[playwright] 
[playwright]   10 failed
[playwright]     ../../tests/theia-explorer-view.test.ts:46:9 › Theia Explorer View › should be possible to close and reopen it 
[playwright]     ../../tests/theia-explorer-view.test.ts:56:9 › Theia Explorer View › should show one folder named "sampleFolder" and one file named "sample.txt" 
[playwright]     ../../tests/theia-main-menu.test.ts:50:9 › Theia Main Menu › should return menu item by name 'New File' 
[playwright]     ../../tests/theia-preference-view.test.ts:30:9 › Preference View › should be visible and active after being opened 
[playwright]     ../../tests/theia-preference-view.test.ts:37:9 › Preference View › should be able to read, set, and reset String preferences 
[playwright]     ../../tests/theia-problems-view.test.ts:30:9 › Theia Problems View › should be visible and active after being opened 
[playwright]     ../../tests/theia-problems-view.test.ts:37:9 › Theia Problems View › should be opened at the bottom and have the title 'Problems' 
[playwright]     ../../tests/theia-quick-command.test.ts:34:9 › Theia Quick Command › should show quick command palette 
[playwright]     ../../tests/theia-quick-command.test.ts:39:9 › Theia Quick Command › should trigger 'About' command after typing 
[playwright]     ../../tests/theia-quick-command.test.ts:53:9 › Theia Quick Command › should trigger 'Toggle Explorer View' command after typing 
[playwright]   10 skipped
[playwright]   28 passed (11m)
```

The runtime was quite long - at least the reported 11 min - and tests that can't be run in a variety of environments are not very productive.

**Reply by JonasHelming** [reply]
I created a ticket to investigate the tests on MacOS, they should of course not be OS specific, we will investigate this.
Independantly of this, I would expect that the uplift of the Monaco editor might break some parts of the page object model (not that many parts), so let us know if you need help fixing that. Please note that the playwright tests currently do not block merging, so if you merge the PR before we fix this, please feel free to assign us a ticket.

**Reply by planger** [reply]
Thanks for bringing that to our awareness!

I'm continuously observing the Playwright tests [in the CI](https://github.com/eclipse-theia/theia/actions/workflows/playwright.yml). So far they were very stable. [This is the first test](https://github.com/eclipse-theia/theia/runs/5321846846?check_suite_focus=true) I saw failing in the CI, which only seems to happen after the Monaco uplift:

```
[playwright]   1) ../../tests/theia-text-editor.test.ts:167:9 › Theia Text Editor › should close without saving =
[playwright] 
[playwright]     Timeout of 60000ms exceeded.
[playwright] 
[playwright]     page.waitForSelector: Target closed
[playwright]     =========================== logs ===========================
[playwright]     waiting for selector "#theia-dialog-shell .dialogBlock .dialogControl button:has-text("Don't save")" to be visible
[playwright]     ============================================================
[playwright] 
[playwright]       170 |         expect(await sampleTextEditor.isDirty()).toBe(true);
[playwright]       171 |
[playwright]     > 172 |         await sampleTextEditor.closeWithoutSave();
[playwright]           |         ^
[playwright]       173 |         expect(await sampleTextEditor.isTabVisible()).toBe(false);
[playwright]       174 |     });
[playwright]       175 |
[playwright] 
[playwright]         at TheiaDialog.buttonElement (/home/runner/work/theia/theia/examples/playwright/src/theia-dialog.ts:59:26)
[playwright]         at TheiaDialog.clickButton (/home/runner/work/theia/theia/examples/playwright/src/theia-dialog.ts:87:42)
[playwright]         at TheiaTextEditor.closeWithoutSave (/home/runner/work/theia/theia/examples/playwright/src/theia-editor.ts:44:26)
[playwright]         at /home/runner/work/theia/theia/examples/playwright/tests/theia-text-editor.test.ts:172:9
```

We are happy to look into it tomorrow. I'm not sure yet whether this is related to the Monaco change or not, but it looks like an adjustment of the page object is needed with the Monaco version change.

> I went to run the tests locally (MacOS) and got numerous failures.

Interesting, that's strange as the tests on master run fine in the CI and on my machine (both Ubuntu though).
In a customer CI, we are running them on Windows and there they are successful too, but we run them on the latest Theia release only.

Can you please attach the full log of your test run? I'll check if I see something.

> The runtime was quite long - at least the reported 11 min - and tests that can't be run in a variety of environments are not very productive.

A successful headless test run usually takes about 2-3mins, depending on the machine. If they are unsuccessful, e.g. because expected elements can't be found, they can take longer since Playwright waits for elements until a timeout is hit, which is 60 secs, I think. So if that happens for several tests, that sums up.

Anyway, the tests certainly should run on a variety of environments. At least they do on Windows and Linux. So we'll have to figure out why they don't seem to run on MacOS. I hope to see something in the full log, once you upload it.

Thanks a lot in advance!

**Reply by colin-grant-work** [reply]
It does seem to have to do with the OS - @vince-fugnitto ran tests in an Ubuntu VM and they passed, but ran them on his Mac and had some failures, and when I switched into an Ubuntu VM, the tests passed again. Not sure what's going wrong on the Mac though.

**Reply by planger** [reply]
Thanks for confirming! I'll work with a colleague in our team who has a Mac to figure it out. It may take a few days though as we might have to ping-pong changes (and I'm on vacation next week). But we'll look into it.

**Reply by planger** [reply]
@colin-grant-work I just debugged into the failing playwright test (`Theia Text Editor > should close without saving`) with your Monaco branch.

The test failure seems to be related to the `Auto Save` preference:
The tests expects that auto save is off and, consequently, the editor remains dirty after changes (which was the default in Theia if I'm not mistaken). Thus, it further expects a dialog to show up that warns the user about unsaved changes. However, auto-save is actually on now (even though it shows off in the menu). Since auto-save is on, the editor isn't dirty, so the dialog never shows up, which eventually causes the test failure.

If I now manually enable and disable auto-save, so that it explicitly is defined in the user settings, the test successfully passes. So it seems to fail only without explicit user settings.

If I can help with anything, please let me know.

**Reply by vince-fugnitto** [reply]
@planger the issue happens on master: https://github.com/eclipse-theia/theia/issues/10793.

**Reply by JonasHelming** [reply]
@vince-fugnitto : I believe there are two issues: 

1. #10793 (failing on Mac) also happening on master. 
2. One specific test failing on the electron uplift, because the auto save behavior seems to have changed (which is actually an interesting catch)

**Reply by planger** [reply]
I think there are two separate issues:

1. The `Theia Text Editor > should close without saving` also fails in the CI but has only been [observed with the Monaco PR](https://github.com/eclipse-theia/theia/runs/5321846846?check_suite_focus=true) so far. I can reproduce this test failure locally, but only with this PR, not on master. Locally it was caused by the Auto-save issue as described [above](https://github.com/eclipse-theia/theia/discussions/10790#discussioncomment-2246786).
2. #10793 happens on master too, but can so far only be reproduced on MacOS and not in the CI

Or did I misunderstand that?

**Reply by planger** [reply]
> @vince-fugnitto : I believe there are two issues:
> 
> 1. [playwright: tests fail on MacOS #10793](https://github.com/eclipse-theia/theia/issues/10793) (failing on Mac) also happening on master.
> 2. One specific test failing on the electron uplift, because the auto save behavior seems to have changed (which is actually an interesting catch)

You were faster @JonasHelming  :)

**Reply by colin-grant-work** [reply]
Re: the Monaco uplift, it is true that the default auto save has changed. VSCode's default differs depending on whether it's running in the browser (in which case autosave is active by default) or in Electron (inactive by default), and I replicated that on the Monaco branch.

An unrelated, and surprising, failure is here: https://github.com/eclipse-theia/theia/runs/5326900058?check_suite_focus=true

I modified the rendering of keybindings, but tests related to the problems widget are failing.

**Reply by planger** [reply]
> An unrelated, and surprising, failure is here: https://github.com/eclipse-theia/theia/runs/5326900058?check_suite_focus=true

This one is indeed surprising. I tried locally but can't reproduce it. I ran the tests 10 times locally with your branch and they didn't fail once. :shrug:

**Reply by planger** [reply]
> Re: the Monaco uplift, it is true that the default auto save has changed. VSCode's default differs depending on whether it's running in the browser (in which case autosave is active by default) or in Electron (inactive by default), and I replicated that on the Monaco branch.

We could explicitly set the auto-save preference before the tests to ensure both tests would run consistently for browser and electron (even though we only run them in the browser for now). If you'd agree, I'm happy to create an issue and schedule it for being addressed by us.

**Comment by JonasHelming** [comment]
@planger Do you think this is actually solved in the meantime?

**Reply by planger** [reply]
Yes, those particular flaky tests have definitely been fixed in the meantime and also in general the builds are pretty stable again thanks to several changes of the past couple of months.
In the past weeks I only saw `theia-explorer-view.test.ts` failing once or twice across all test runs. Aside from those everything was constantly green over quite some time. We'll continue observing, but at the moment I'd say there is no particular flakiness.

Edit: PS: This comment is in the context of Playwright

**Comment by JonasHelming** [comment, marked as answer]
@colin-grant-work Are you fine with setting this to "answered"?
